arch/x86/kvm/pkvm/pkvm.c

```diff
@@ -195,6 +196,219 @@ static void pkvm_vm_destroy(int vm_handle, struct pkvm_memcache *mc)
 	teardown_donated_memory(mc, (void *)pkvm_vm, pkvm_vm->size);
 }
 
+static int attach_pkvm_vcpu_to_vm(struct pkvm_vm *pkvm_vm, struct pkvm_vcpu *pkvm_vcpu)
+{
+	struct kvm *kvm = &pkvm_vm->kvm;
+	int vcpu_handle;
+
+	pkvm_spin_lock(&pkvm_vm->lock);
+
+	if (kvm->created_vcpus == KVM_MAX_VCPUS) {
+		pkvm_spin_unlock(&pkvm_vm->lock);
+		return -EINVAL;
+	}
+	vcpu_handle = kvm->created_vcpus++;
+	pkvm_vcpu->vcpu.arch.pkvm_vcpu_handle = vcpu_handle;
+	pkvm_vcpu->pkvm_vm = pkvm_vm;
+	pkvm_vm->vcpus[vcpu_handle] = pkvm_vcpu;
+
+	pkvm_spin_unlock(&pkvm_vm->lock);
+
+	atomic_set(&pkvm_vm->vcpu_refs[vcpu_handle], 1);
+
+	return vcpu_handle;
+}
+
+static int setup_vcpu_lapic(struct kvm_vcpu *vcpu)
+{
+	struct pkvm_vcpu *pkvm_vcpu = to_pkvm_vcpu(vcpu);
+	struct kvm_lapic *shared_apic, *apic;
+	struct kvm_vcpu *shared_vcpu;
+	void *apic_regs = NULL;
+	phys_addr_t apic_pa;
+	size_t apic_size;
+	int ret;
+
+	shared_vcpu = pkvm_vcpu->shared_vcpu;
+	if (!lapic_in_kernel(pkvm_vcpu->shared_vcpu))
+		return 0;
+
+	/*
+	 * The host allocates contiguous memory pages for the pkvm_vcpu
+	 * structure, including the struct kvm_lapic memory page if lapic
+	 * is emulated.
+	 */
+	apic_pa = __pkvm_pa(pkvm_vcpu) + pkvm_vcpu->size;
+	apic_size = PAGE_ALIGN(sizeof(struct kvm_lapic));
+	ret = pkvm_host_donate_hyp(apic_pa, apic_size, true);
+	if (ret)
+		return ret;
+
+	/*
+	 * Increase pkvm_vcpu->size with the size of donated struct kvm_lapic
+	 * memory pages.
+	 */
+	pkvm_vcpu->size += apic_size;
+	apic = __pkvm_va(apic_pa);
+
+	shared_apic = kern_pkvm_va(shared_vcpu->arch.apic);
+	/*
+	 * Temporary sharing host's apic page to access its elements for setting
+	 * up pKVM's apic page. It will be unshared after that.
+	 */
+	ret = pkvm_host_share_hyp(__pkvm_pa(shared_apic), apic_size);
+	if (ret)
+		goto undonate_apic;
+
+	apic_regs = kern_pkvm_va(shared_apic->regs);
+	if (!apic_regs) {
+		ret = -EINVAL;
+		goto unshare_apic;
+	}
+
+	ret = pkvm_host_share_hyp(__pkvm_pa(apic_regs), PAGE_SIZE);
+	if (ret)
+		goto unshare_apic;
+
+	apic->regs = apic_regs;
+	apic->apicv_active = shared_apic->apicv_active;
+	apic->nr_lvt_entries = kvm_apic_calc_nr_lvt_entries(vcpu);
+	apic->vcpu = vcpu;
+
+	pkvm_host_unshare_hyp(__pkvm_pa(shared_apic), apic_size);
+
+	vcpu->arch.apic = apic;
+
+	return 0;
+
+unshare_apic:
+	pkvm_host_unshare_hyp(__pkvm_pa(shared_apic), apic_size);
+undonate_apic:
+	pkvm_hyp_donate_host(apic_pa, apic_size, false);
+	pkvm_vcpu->size -= apic_size;
+	return ret;
+}
+
+static void unsetup_vcpu_lapic(struct kvm_vcpu *vcpu)
+{
+	size_t apic_size = PAGE_ALIGN(sizeof(struct kvm_lapic));
+	struct kvm_lapic *apic = xchg(&vcpu->arch.apic, NULL);
```

Comment:
Do we need to reset vcpu->arch.apic to NULL here? We only use unsetup_vcpu_lapic() when the vCPU is about to be freed anyway?

Reply:
Previously, the struct kvm_lapic memory page was undonated in this function. From concept point of view, apic should not point to this page anymore. Now with making struct kvm_lapic as part of the same page-aligned donated region, it is not necessary to reset vcpu->arch.apic as NULL.